File: experiments/verification-poc/lrat.py
#!/usr/bin/env python3
"""Independent ASCII LRAT RUP/deletion checker (CaDiCaL 2.1.3 subset).

Negative RAT hints and binary LRAT are explicitly unsupported. This module has
no dependency on Oak, the CNF encoder, or the proof producer.
"""
import argparse
import json
from pathlib import Path
import sys


class InvalidProof(ValueError):
    pass


def need(ok, why):
    if not ok:
        raise InvalidProof(why)


def dimacs(text):
    header = None
    data = []
    for line in text.splitlines():
        line = line.strip()
        if not line or line.startswith('c'):
            continue
        if line.startswith('p'):
            need(header is None and not data, 'duplicate or misplaced DIMACS header')
            words = line.split()
            need(len(words) == 4 and words[:2] == ['p', 'cnf'], 'invalid DIMACS header')
            header = tuple(map(int, words[2:]))
            need(all(x >= 0 for x in header), 'negative DIMACS counts')
        else:
            need(header is not None, 'missing DIMACS header')
            data.extend(map(int, line.split()))
    need(header is not None, 'missing DIMACS header')
    clauses, clause = [], []
    for lit in data:
        if lit == 0:
            clauses.append(frozenset(clause)); clause = []
        else:
            need(abs(lit) <= header[0], 'literal exceeds declared variable domain')
            clause.append(lit)
    need(not clause and (not data or data[-1] == 0), 'unterminated clause')
    need(len(clauses) == header[1], 'clause count mismatch')
    return header[0], clauses


def rup(database, clause, hints):
    # Assumed TRUE literals are precisely the negated proposed clause.
    assigned = {-lit for lit in clause}
    for hint in hints:
        need(hint > 0, 'RAT/negative hints unsupported; use ASCII RUP LRAT')
        need(hint in database, 'hint refers to absent/deleted/future clause')
    if any(-lit in assigned for lit in assigned):
        return  # tautological proposed clause
    for position, hint in enumerate(hints):
        premise = database[hint]
        need(not any(lit in assigned for lit in premise), 'hint clause is already satisfied')
        remaining = premise - {-lit for lit in assigned}
        if not remaining:
            need(position == len(hints) - 1, 'unused hints after conflict')
            return
        need(len(remaining) == 1, 'hint is neither unit nor conflicting')
        assigned.add(next(iter(remaining)))
    raise InvalidProof('RUP chain did not derive a conflict')


def check(cnf_text, proof_text):
    variables, clauses = dimacs(cnf_text)
    database = {i + 1: c for i, c in enumerate(clauses)}
    last_added = len(database)
    has_empty = any(not c for c in clauses)
    additions = deletions = 0
    for number, line in enumerate(proof_text.splitlines(), 1):
        words = line.split()
        if not words or words[0] == 'c':
            continue
        try:
            need(len(words) >= 3, 'short LRAT line')
            identifier = int(words[0])
            if words[1] == 'd':
                ids = list(map(int, words[2:]))
                need(identifier >= last_added and ids[-1] == 0 and all(x > 0 for x in ids[:-1]), 'invalid deletion line')
                need(len(ids[:-1]) == len(set(ids[:-1])), 'duplicate deletion')
                for removed in ids[:-1]:
                    need(removed in database, 'deletion refers to absent clause')
                    del database[removed]
                    deletions += 1
                continue
            need(identifier > last_added, 'addition ids must strictly increase')
            values = list(map(int, words[1:]))
            need(values.count(0) == 2 and values[-1] == 0, 'expected clause and hints terminated by zero')
            split = values.index(0)
            clause, hints = frozenset(values[:split]), values[split + 1:-1]
            need(all(abs(x) <= variables for x in clause), 'proof uses an undeclared variable')
            rup(database, clause, hints)
            database[identifier] = clause
            last_added = identifier
            additions += 1
            has_empty |= not clause
        except (ValueError, IndexError) as error:
            raise InvalidProof(f'LRAT line {number}: {error}') from error
    need(has_empty, 'proof never establishes the empty clause')
    return {'accepted': True, 'claim': 'CNF unsatisfiable', 'format': 'ASCII LRAT RUP/deletion', 'additions': additions, 'deletions': deletions}


def main():
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument('cnf', type=Path); p.add_argument('proof', type=Path)
    a = p.parse_args()
    try:
        need(a.cnf.stat().st_size < 20_000_000 and a.proof.stat().st_size < 50_000_000, 'prototype input size limit')
        print(json.dumps(check(a.cnf.read_text(), a.proof.read_text()), indent=2))
        return 0
    except (ValueError, OSError) as error:
        print(json.dumps({'accepted': False, 'reason': str(error)})); return 1


if __name__ == '__main__':
    sys.exit(main())

File: experiments/verification-poc/internal/lrat/lrat.go
// Package lrat checks ASCII LRAT RUP additions and deletions. It has no Oak,
// encoder, producer, or external solver dependency. RAT hints are unsupported.
package lrat

import (
	"fmt"
	"strconv"
	"strings"
)

type Clause map[int]bool
type Formula struct {
	Variables int
	Clauses   []Clause
}
type Result struct {
	Accepted  bool `json:"accepted"`
	Additions int  `json:"additions"`
	Deletions int  `json:"deletions"`
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
func integer(s string) (int, error) {
	n, err := strconv.Atoi(s)
	if err != nil || n < -2147483647 || n > 2147483647 {
		return 0, fmt.Errorf("invalid bounded integer %q", s)
	}
	return n, nil
}
func Parse(text string) (Formula, error) {
	var f Formula
	if len(text) > 20_000_000 {
		return f, fmt.Errorf("CNF input limit")
	}
	header := false
	expected := 0
	clause := Clause{}
	for _, line := range strings.Split(text, "\n") {
		words := strings.Fields(line)
		if len(words) == 0 || words[0] == "c" {
			continue
		}
		if words[0] == "p" {
			if header || len(words) != 4 || words[1] != "cnf" {
				return f, fmt.Errorf("invalid DIMACS header")
			}
			n, e := integer(words[2])
			if e != nil || n < 0 {
				return f, fmt.Errorf("invalid variable count")
			}
			m, e := integer(words[3])
			if e != nil || m < 0 {
				return f, fmt.Errorf("invalid clause count")
			}
			f.Variables = n
			expected = m
			header = true
			continue
		}
		if !header {
			return f, fmt.Errorf("missing DIMACS header")
		}
		for _, word := range words {
			n, e := integer(word)
			if e != nil {
				return f, e
			}
			if n == 0 {
				f.Clauses = append(f.Clauses, clause)
				clause = Clause{}
			} else {
				if abs(n) > f.Variables {
					return f, fmt.Errorf("literal outside variable domain")
				}
				clause[n] = true
			}
		}
	}
	if !header || len(clause) != 0 || len(f.Clauses) != expected {
		return f, fmt.Errorf("unterminated clause or count mismatch")
	}
	return f, nil
}
func rup(db map[int]Clause, c Clause, hints []int) error {
	assigned := Clause{}
	for x := range c {
		assigned[-x] = true
	}
	for _, id := range hints {
		if id <= 0 || db[id] == nil {
			return fmt.Errorf("unsupported RAT or absent/deleted/future hint")
		}
	}
	for x := range assigned {
		if assigned[-x] {
			return nil
		}
	}
	for i, id := range hints {
		remaining := 0
		unit := 0
		for x := range db[id] {
			if assigned[x] {
				return fmt.Errorf("hint is already satisfied")
			}
			if !assigned[-x] {
				remaining++
				unit = x
			}
		}
		if remaining == 0 {
			if i != len(hints)-1 {
				return fmt.Errorf("unused hints after conflict")
			}
			return nil
		}
		if remaining != 1 {
			return fmt.Errorf("hint is not unit or conflicting")
		}
		assigned[unit] = true
	}
	return fmt.Errorf("RUP chain has no conflict")
}
func Check(cnf, proof string) (Result, error) {
	result := Result{}
	if len(proof) > 50_000_000 {
		return result, fmt.Errorf("proof input limit")
	}
	f, err := Parse(cnf)
	if err != nil {
		return result, err
	}
	db := map[int]Clause{}
	empty := false
	for i, c := range f.Clauses {
		db[i+1] = c
		if len(c) == 0 {
			empty = true
		}
	}
	last := len(db)
	for index, line := range strings.Split(proof, "\n") {
		w := strings.Fields(line)
		if len(w) == 0 || w[0] == "c" {
			continue
		}
		fail := func(s string) (Result, error) { return result, fmt.Errorf("LRAT line %d: %s", index+1, s) }
		if len(w) < 3 {
			return fail("short line")
		}
		id, e := integer(w[0])
		if e != nil {
			return fail(e.Error())
		}
		if w[1] == "d" {
			if id < last || w[len(w)-1] != "0" {
				return fail("invalid deletion")
			}
			seen := map[int]bool{}
			for _, word := range w[2 : len(w)-1] {
				n, e := integer(word)
				if e != nil || n <= 0 || db[n] == nil || seen[n] {
					return fail("invalid deletion reference")
				}
				seen[n] = true
				delete(db, n)
				result.Deletions++
			}
			continue
		}
		if id <= last {
			return fail("addition IDs must increase")
		}
		nums := []int{}
		zeros := []int{}
		for _, word := range w[1:] {
			n, e := integer(word)
			if e != nil {
				return fail(e.Error())
			}
			if n == 0 {
				zeros = append(zeros, len(nums))
			}
			nums = append(nums, n)
		}
		if len(zeros) != 2 || zeros[1] != len(nums)-1 {
			return fail("expected two zero terminators")
		}
		c := Clause{}
		for _, x := range nums[:zeros[0]] {
			if abs(x) > f.Variables {
				return fail("undeclared variable")
			}
			c[x] = true
		}
		if e := rup(db, c, nums[zeros[0]+1:len(nums)-1]); e != nil {
			return fail(e.Error())
		}
		db[id] = c
		last = id
		result.Additions++
		if len(c) == 0 {
			empty = true
		}
	}
	if !empty {
		return result, fmt.Errorf("proof never establishes the empty clause")
	}
	result.Accepted = true
	return result, nil
}

// CheckRUPDecoded exposes one decoded RUP step for differential testing of
// bounded self-hosted kernels. Clauses and literals use DIMACS conventions.
func CheckRUPDecoded(variables int, clauses [][]int, target []int, hints []int) error {
	if variables <= 0 {
		return fmt.Errorf("invalid variable count")
	}
	db := map[int]Clause{}
	convert := func(xs []int) (Clause, error) {
		c := Clause{}
		for _, x := range xs {
			if x == 0 || abs(x) > variables {
				return nil, fmt.Errorf("literal outside variable domain")
			}
			c[x] = true
		}
		return c, nil
	}
	for i, xs := range clauses {
		c, e := convert(xs)
		if e != nil {
			return e
		}
		db[i+1] = c
	}
	c, e := convert(target)
	if e != nil {
		return e
	}
	return rup(db, c, hints)
}
